go/database/mpt/decoder_test.go

```diff
@@ -0,0 +1,374 @@
+		"two items node, path is list":             rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{list, str}}),
+		"possible value but nested list":           rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{rlp.String{Str: []byte{0x31, 0x23, 0x45}}, rlp.List{}}}),
+		"possible value but too long key":          rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{rlp.String{Str: append([]byte{0x31, 0x23, 0x45}, strLongerThan32.Str...)}, str}}),
+		"possible ext but nested list":             rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{rlp.String{Str: []byte{0x00, 0x12, 0x34}}, rlp.List{}}}),
+		"possible ext but too long hash":           rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{rlp.String{Str: []byte{0x00, 0x12, 0x34}}, strLongerThan32}}),
+		"possible account but 3 items nested list": rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, threeItemsList}}),
+		"possible account but nested empty":        rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, rlp.String{Str: trailingBytes}}}),
+		"possible account but nested not list":     rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, nonListSubItem}}),
+		"possible account but nonce not string":    rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, nonceNotStr}}),
+		"possible account but nonce wrong number":  rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, nonceWrongNumber}}),
+		"possible account but balance not string":  rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, balanceNotStr}}),
+		"possible account but balance too long":    rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, balanceWrongNumber}}),
+		"possible account but storage not string":  rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, storageNotStr}}),
+		"possible account but storage too long":    rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, storageTooLong}}),
+		"possible account but codeHash not string": rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, codeHashNotStr}}),
+		"possible account but codeHash too long":   rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{str, codeHashTooLong}}),
+		"possible account but long address":        rlp.EncodeInto([]byte{}, rlp.List{Items: []rlp.Item{tooLongNumberItemsList, strLongerThan32}}),
+		"possible branch too long child":           rlp.EncodeInto([]byte{}, rlp.List{Items: childrenTooLongHashes}),
+		"possible branch child not strings":        rlp.EncodeInto([]byte{}, rlp.List{Items: childrenNotStrings}),
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			if _, err := DecodeFromRlp(test); err == nil {
+				t.Fatalf("expected error, got nil")
+			}
+		})
+	}
+}
+
+func Test_isCompactPathLeafNode(t *testing.T) {
+	tests := []struct {
+		path   []byte
+		isLeaf bool
+	}{
+		{[]byte{0b_0000_0000}, false},
+		{[]byte{0b_0001_0000}, false},
+		{[]byte{0b_0010_0000}, true},
+		{[]byte{0b_0011_0000}, true},
```

Comment:
nit: if you want to stress-test this, test all possible 256 values for the first byte.

Reply:
good idea, updated